README.md
# gsoc-match

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2017 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-03-03
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-03-03
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/gsoc-match/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/gsoc-match/compare/0.1.0...0.1.1


test/gsoc_match/core_test.clj
(ns gsoc-match.core-test
  (:require [clojure.test :refer :all]
            [gsoc-match.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


project.clj
(defproject org.clojure/core.match "0.3.0-alpha4-SNAPSHOT"
  :description "Optimized pattern matching for Clojure"

  :jvm-opts ^:replace ["-Xmx512m" "-server"]

  :test-paths ["src/test/clojure"]
  :source-paths ["src/main/clojure"]

  :dependencies [[org.clojure/clojure "1.8.0"]
                 [org.clojure/tools.analyzer.jvm "0.6.5"]
                 [org.clojure/clojurescript "0.0-2496" :scope "provided"]
                 [org.clojure/core.match "0.3.0-alpha4"]]

  :plugins [[lein-cljsbuild "1.0.4-SNAPSHOT"]
            [cider/cider-nrepl "0.14.0"]]

  :cljsbuild
  {:builds
   [{:id "test"
     :source-paths ["src/test/cljs"]
     :compiler {:output-to "out/test.js"
                :static-fns true
                :optimizations :simple}}
    {:id "test-adv"
     :source-paths ["src/test/cljs"]
     :compiler {:output-to "out/test.js"
                :optimizations :advanced}}]})


src/gsoc_match/core.clj
(ns gsoc-match.core
  (:require [clojure.core.match :refer [match]]))



(let [v [1 2]]
  (match [v]
    [] [3 1] :a0
    [([1 a] :as b)] [:a1 a b]))








(let [y '(+ 1 3)
      z true]
  (match [(+ 1 3) y z]
         [false ([1] :seq) true ] 1
         [false true _ ] 2))

(let*
 [y '(+ 1 3) z true]
 (let*
  [ocr-15480 (+ 1 3) y y z z]
  (try
   (if
    (clojure.core/= ocr-15480 false)
    (try
     (if
      (let*
       [and__4467__auto__
        (let*
         [or__4469__auto__ (clojure.core/seq? y)]
         (if
          or__4469__auto__
          or__4469__auto__
          (clojure.core/sequential? y)))]
       (if and__4467__auto__ (clojure.core/seq y) and__4467__auto__))
      (try
       (let*
        [y_head__15485 (clojure.core/first y)]
        (if
         (clojure.core/= y_head__15485 1)
         (try
       ff   (let*
           [y_tail__15486 (clojure.core/rest y)]
           (if
            (clojure.core/empty? y_tail__15486)
            (try
             (if
              (clojure.core/= z true)
              1
              (if :else (throw clojure.core.match/backtrack) nil))
             (catch
              Exception
              e__14590__auto__
              (if
               (clojure.core/identical?
                e__14590__auto__
                clojure.core.match/backtrack
)
               (do (throw clojure.core.match/backtrack))
               (throw e__14590__auto__))))
            (if :else (throw clojure.core.match/backtrack) nil)))
          (catch
           Exception
           e__14590__auto__
           (if
            (clojure.core/identical?
             e__14590__auto__
             clojure.core.match/backtrack)
            (do (throw clojure.core.match/backtrack))
            (throw e__14590__auto__))))
         (if :else (throw clojure.core.match/backtrack) nil)))
       (catch
        Exception
        e__14590__auto__
        (if
         (clojure.core/identical?
          e__14590__auto__
          clojure.core.match/backtrack)
         (do (throw clojure.core.match/backtrack))
         (throw e__14590__auto__))))
      (if :else (throw clojure.core.match/backtrack) nil))
     (catch
      Exception
      e__14590__auto__
      (if
       (clojure.core/identical?
        e__14590__auto__
        clojure.core.match/backtrack)
       (do
        (try
         (if

          (clojure.core/= y true)
          2
          (if :else (throw clojure.core.match/backtrack) nil))
         (catch
          Exception
          e__14590__auto__
          (if
           (clojure.core/identical?
            e__14590__auto__
            clojure.core.match/backtrack)
           (do (throw clojure.core.match/backtrack))
           (throw e__14590__auto__)))))
       (throw e__14590__auto__))))
    (if :else (throw clojure.core.match/backtrack) nil))
   (catch
    Exception
    e__14590__auto__
    (if
     (clojure.core/identical?
      e__14590__auto__
      clojure.core.match/backtrack)
     (do
      (throw
       (new
        java.lang.IllegalArgumentException
        (clojure.core/str
         "No matching clause: "
         ocr-15480
         " "
         y
         " "
         z))))
     (throw e__14590__auto__))))))






(let*
 [ocr-15543 'x ocr-15544 'y ocr-15545 'z]
 (try
  (clojure.core/cond
   (clojure.core/= ocr-15544 false)
   (try
    (clojure.core/cond
     (clojure.core/= ocr-15545 true)
     1
     :else
     (throw clojure.core.match/backtrack))
    (catch
     Exception
     e__14592__auto__
     (if
      (clojure.core/identical?
       e__14592__auto__
       clojure.core.match/backtrack)
      (do (throw clojure.core.match/backtrack))
      (throw e__14592__auto__))))
   :else
   (throw clojure.core.match/backtrack))
  (catch
   Exception
   e__14592__auto__
   (if
    (clojure.core/identical?
     e__14592__auto__
     clojure.core.match/backtrack)
    (do
     (try
      (clojure.core/cond
       (clojure.core/= ocr-15544 true)
       (try
        (clojure.core/cond
         (clojure.core/= ocr-15543 false)
         2
         :else
         (throw clojure.core.match/backtrack))
        (catch
         Exception
         e__14592__auto__
         (if
          (clojure.core/identical?
           e__14592__auto__

           clojure.core.match/backtrack)
          (do (throw clojure.core.match/backtrack))
          (throw e__14592__auto__))))
       :else
       (throw clojure.core.match/backtrack))
      (catch
       Exception
       e__14592__auto__
       (if
        (clojure.core/identical?
         e__14592__auto__
         clojure.core.match/backtrack)
        (do
         (try
          (clojure.core/cond
           (clojure.core/= ocr-15545 false)
           3
           :else
           (throw clojure.core.match/backtrack))
          (catch
           Exception
           e__14592__auto__
           (if
            (clojure.core/identical?
             e__14592__auto__
             clojure.core.match/backtrack)
            (do
             (try
              (clojure.core/cond
               (clojure.core/= ocr-15545 true)
               4
               :else
               (throw clojure.core.match/backtrack))
              (catch
               Exception
               e__14592__auto__
               (if
                (clojure.core/identical?
                 e__14592__auto__
                 clojure.core.match/backtrack)
                (do
                 (throw
                  (java.lang.IllegalArgumentException.
                   (clojure.core/str
                    "No matching clause: "
                    ocr-15543
                    " "
                    ocr-15544
                    " "
                    ocr-15545))))
                (throw e__14592__auto__)))))
            (throw e__14592__auto__)))))
        (throw e__14592__auto__)))))
    (throw e__14592__auto__)))))











;------------------------------




(let*
 [ocr-15543 'x ocr-15544 'y ocr-15545 'z]
 (try
  (clojure.core/cond
   (clojure.core/= y false)
   (try
    (clojure.core/cond
     (clojure.core/= z true)
     1
     :else
     (throw clojure.core.match/backtrack))
    (catch
     Exception
     e__14592__auto__
     (if
      (clojure.core/identical?
       e__14592__auto__
       clojure.core.match/backtrack)
      (do (throw clojure.core.match/backtrack))
      (throw e__14592__auto__))))
   :else
   (throw clojure.core.match/backtrack))
  (catch
   Exception
   e__14592__auto__
   (if
    (clojure.core/identical?
     e__14592__auto__
     clojure.core.match/backtrack)
    (do
     (try
      (clojure.core/cond
       (clojure.core/= y true)
       (try
        (clojure.core/cond
         (clojure.core/= x false)
         2
         :else
         (throw clojure.core.match/backtrack))
        (catch
         Exception
         e__14592__auto__
         (if
          (clojure.core/identical?
           e__14592__auto__

           clojure.core.match/backtrack)
          (do (throw clojure.core.match/backtrack))
          (throw e__14592__auto__))))
       :else
       (throw clojure.core.match/backtrack))
      (catch
       Exception
       e__14592__auto__
       (if
        (clojure.core/identical?
         e__14592__auto__
         clojure.core.match/backtrack)
        (do
         (try
          (clojure.core/cond
           (clojure.core/= z false)
           3
           :else
           (throw clojure.core.match/backtrack))
          (catch
           Exception
           e__14592__auto__
           (if
            (clojure.core/identical?
             e__14592__auto__
             clojure.core.match/backtrack)
            (do
             (try
              (clojure.core/cond
               (clojure.core/= z true)
               4
               :else
               (throw clojure.core.match/backtrack))
              (catch
               Exception
               e__14592__auto__
               (if
                (clojure.core/identical?
                 e__14592__auto__
                 clojure.core.match/backtrack)
                (do
                 (throw
                  (java.lang.IllegalArgumentException.
                   (clojure.core/str
                    "No matching clause: "
                    x
                    " "
                    ocr-15544
                    " "
                    z))))
                (throw e__14592__auto__)))))
            (throw e__14592__auto__)))))
        (throw e__14592__auto__)))))
    (throw e__14592__auto__)))))
















[:a1 2 [1 2]]


(defn foo
  "I don't do a whole lot."
  [x]
  (println x "Hello, World!"))



src/gsoc_match/core2.clj

(try
  (cond
    (= y false) (try
                  (cond (= z true) 1
                        :else (throw backtrack))
                  (catch Exception e
                    (if (identical? e backtrack)
                      (throw backtrack)
                      (throw e))))
    
    :else (throw backtrack))
  (catch Exception e
    (if (identical? e backtrack)
      (do
        (try
          (cond
            (= y true) (try (cond
                              (= x false) 2
                              :else (throw backtrack))
                            (catch Exception e
                              (if (identical? e backtrack)
                                (throw backtrack)
                                (throw e))))
            :else (throw backtrack))
          (catch Exception e
            (if (identical? e backtrack)
              (do (try
                    (cond
                      (= z false) 3
                      :else (trhow backtrack))
                    (catch  Exception e
                      (if (identical? e backtrack)
                        (do (try
                              (cond
                                (= z true) 4
                                :else (throw backtrack))
                              (catch Exception e
                                (if (identical? e backtrack)
                                  (throw no-matching-e)
                                  (throw e)))))
                        (throw e)))))
              (throw e)))))
      (throw e))))

;; this should be

(cond (= y false)  (cond (= z true) 1)
      (= y true) (cond (= x false) 2
                       (= z false) 3
                       (= z true) 4))



; loop + jump table (via case) | not cond? 








src/gsoc_match/scratch2.clj
(ns gsoc-match.core
  (:require [clojure.core.match :refer [match]]))



(let [v [1 2]]
  (match [v]
    [] [3 1] :a0
    [([1 a] :as b)] [:a1 a b]))








(let [y '(+ 1 3)
      z true]
  (match [(+ 1 3) y z]
         [false ([1] :seq) true ] 1
         [true true _ ] 2))

(let*
 [y '(+ 1 3) z true]
 (let*
  [ocr-15480 (+ 1 3) y y z z]
  (try
   (if
    (clojure.core/= ocr-15480 false)
    (try
     (if
      (let*
       [and__4467__auto__
        (let*
         [or__4469__auto__ (clojure.core/seq? y)]
         (if
          or__4469__auto__
          or__4469__auto__
          (clojure.core/sequential? y)))]
       (if and__4467__auto__ (clojure.core/seq y) and__4467__auto__))
      (try
       (let*
        [y_head__15485 (clojure.core/first y)]
        (if
         (clojure.core/= y_head__15485 1)
         (try
       ff   (let*
           [y_tail__15486 (clojure.core/rest y)]
           (if
            (clojure.core/empty? y_tail__15486)
            (try
             (if
              (clojure.core/= z true)
              1
              (if :else (throw clojure.core.match/backtrack) nil))
             (catch
              Exception
              e__14590__auto__
              (if
               (clojure.core/identical?
                e__14590__auto__
                clojure.core.match/backtrack
)
               (do (throw clojure.core.match/backtrack))
               (throw e__14590__auto__))))
            (if :else (throw clojure.core.match/backtrack) nil)))
          (catch
           Exception
           e__14590__auto__
           (if
            (clojure.core/identical?
             e__14590__auto__
             clojure.core.match/backtrack)
            (do (throw clojure.core.match/backtrack))
            (throw e__14590__auto__))))
         (if :else (throw clojure.core.match/backtrack) nil)))
       (catch
        Exception
        e__14590__auto__
        (if
         (clojure.core/identical?
          e__14590__auto__
          clojure.core.match/backtrack)
         (do (throw clojure.core.match/backtrack))
         (throw e__14590__auto__))))
      (if :else (throw clojure.core.match/backtrack) nil))
     (catch
      Exception
      e__14590__auto__
      (if
       (clojure.core/identical?
        e__14590__auto__
        clojure.core.match/backtrack)
       (do
        (try
         (if

          (clojure.core/= y true)
          2
          (if :else (throw clojure.core.match/backtrack) nil))
         (catch
          Exception
          e__14590__auto__
          (if
           (clojure.core/identical?
            e__14590__auto__
            clojure.core.match/backtrack)
           (do (throw clojure.core.match/backtrack))
           (throw e__14590__auto__)))))
       (throw e__14590__auto__))))
    (if :else (throw clojure.core.match/backtrack) nil))
   (catch
    Exception
    e__14590__auto__
    (if
     (clojure.core/identical?
      e__14590__auto__
      clojure.core.match/backtrack)
     (do
      (throw
       (new
        java.lang.IllegalArgumentException
        (clojure.core/str
         "No matching clause: "
         ocr-15480
         " "
         y
         " "
         z))))
     (throw e__14590__auto__))))))






(let*
 [ocr-15543 'x ocr-15544 'y ocr-15545 'z]
 (try
  (clojure.core/cond
   (clojure.core/= ocr-15544 false)
   (try
    (clojure.core/cond
     (clojure.core/= ocr-15545 true)
     1
     :else
     (throw clojure.core.match/backtrack))
    (catch
     Exception
     e__14592__auto__
     (if
      (clojure.core/identical?
       e__14592__auto__
       clojure.core.match/backtrack)
      (do (throw clojure.core.match/backtrack))
      (throw e__14592__auto__))))
   :else
   (throw clojure.core.match/backtrack))
  (catch
   Exception
   e__14592__auto__
   (if
    (clojure.core/identical?
     e__14592__auto__
     clojure.core.match/backtrack)
    (do
     (try
      (clojure.core/cond
       (clojure.core/= ocr-15544 true)
       (try
        (clojure.core/cond
         (clojure.core/= ocr-15543 false)
         2
         :else
         (throw clojure.core.match/backtrack))
        (catch
         Exception
         e__14592__auto__
         (if
          (clojure.core/identical?
           e__14592__auto__

           clojure.core.match/backtrack)
          (do (throw clojure.core.match/backtrack))
          (throw e__14592__auto__))))
       :else
       (throw clojure.core.match/backtrack))
      (catch
       Exception
       e__14592__auto__
       (if
        (clojure.core/identical?
         e__14592__auto__
         clojure.core.match/backtrack)
        (do
         (try
          (clojure.core/cond
           (clojure.core/= ocr-15545 false)
           3
           :else
           (throw clojure.core.match/backtrack))
          (catch
           Exception
           e__14592__auto__
           (if
            (clojure.core/identical?
             e__14592__auto__
             clojure.core.match/backtrack)
            (do
             (try
              (clojure.core/cond
               (clojure.core/= ocr-15545 true)
               4
               :else
               (throw clojure.core.match/backtrack))
              (catch
               Exception
               e__14592__auto__
               (if
                (clojure.core/identical?
                 e__14592__auto__
                 clojure.core.match/backtrack)
                (do
                 (throw
                  (java.lang.IllegalArgumentException.
                   (clojure.core/str
                    "No matching clause: "
                    ocr-15543
                    " "
                    ocr-15544
                    " "
                    ocr-15545))))
                (throw e__14592__auto__)))))
            (throw e__14592__auto__)))))
        (throw e__14592__auto__)))))
    (throw e__14592__auto__)))))











;------------------------------




(let*
 [ocr-15543 'x ocr-15544 'y ocr-15545 'z]
 (try
  (clojure.core/cond
   (clojure.core/= y false)
   (try
    (clojure.core/cond
     (clojure.core/= z true)
     1
     :else
     (throw clojure.core.match/backtrack))
    (catch
     Exception
     e__14592__auto__
     (if
      (clojure.core/identical?
       e__14592__auto__
       clojure.core.match/backtrack)
      (do (throw clojure.core.match/backtrack))
      (throw e__14592__auto__))))
   :else
   (throw clojure.core.match/backtrack))
  (catch
   Exception
   e__14592__auto__
   (if
    (clojure.core/identical?
     e__14592__auto__
     clojure.core.match/backtrack)
    (do
     (try
      (clojure.core/cond
       (clojure.core/= y true)
       (try
        (clojure.core/cond
         (clojure.core/= x false)
         2
         :else
         (throw clojure.core.match/backtrack))
        (catch
         Exception
         e__14592__auto__
         (if
          (clojure.core/identical?
           e__14592__auto__

           clojure.core.match/backtrack)
          (do (throw clojure.core.match/backtrack))
          (throw e__14592__auto__))))
       :else
       (throw clojure.core.match/backtrack))
      (catch
       Exception
       e__14592__auto__
       (if
        (clojure.core/identical?
         e__14592__auto__
         clojure.core.match/backtrack)
        (do
         (try
          (clojure.core/cond
           (clojure.core/= z false)
           3
           :else
           (throw clojure.core.match/backtrack))
          (catch
           Exception
           e__14592__auto__
           (if
            (clojure.core/identical?
             e__14592__auto__
             clojure.core.match/backtrack)
            (do
             (try
              (clojure.core/cond
               (clojure.core/= z true)
               4
               :else
               (throw clojure.core.match/backtrack))
              (catch
               Exception
               e__14592__auto__
               (if
                (clojure.core/identical?
                 e__14592__auto__
                 clojure.core.match/backtrack)
                (do
                 (throw
                  (java.lang.IllegalArgumentException.
                   (clojure.core/str
                    "No matching clause: "
                    x
                    " "
                    ocr-15544
                    " "
                    z))))
                (throw e__14592__auto__)))))
            (throw e__14592__auto__)))))
        (throw e__14592__auto__)))))
    (throw e__14592__auto__)))))
















[:a1 2 [1 2]]


(defn foo
  "I don't do a whole lot."
  [x]
  (println x "Hello, World!"))


src/main/clojure/clojure/core/match/array.clj
;; WARNING: this namespace is experimental
(ns ^{:skip-wiki true}
  clojure.core.match.array
  (:refer-clojure :exclude [compile])
  (:use [clojure.core.match :as m]))

(set! *warn-on-reflection* true)

;; =============================================================================
;; Shared

(derive ::m/array ::m/vector)
(defmethod nth-inline ::m/array
  [t ocr i]
  `(aget ~ocr ~i))
(defmethod count-inline ::m/array
  [t ocr]
  `(alength ~ocr))
(defmethod subvec-inline ::m/array
  ([_ ocr start] ocr)
  ([_ ocr start end] ocr))

;; =============================================================================
;; ints

(derive ::m/ints ::m/array)
(defmethod tag ::m/ints
  [_] "[I")

;; =============================================================================
;; objects

(derive ::m/objects ::m/array)
(defmethod tag ::m/objects
  [_] "[Ljava.lang.Object;")

(comment
  ;; specialize based on type hints in match
  (let [x (int-array [1 2 3])]
    (match [^ints x]
      [[_ _ 2]] :a0
      [[1 1 3]] :a1
      [[1 2 3]] :a2
      :else :a3))

  ;; FIXME
  (let [x (int-array [1 2 3 4])]
    (match [^ints x]
      [[_ _ 2 & _]] :a0
      [[1 1 3 & _]] :a1
      [[1 2 3 & _]] :a2
      :else :a3))
  
  (let [x (int-array [1 2 3])]
    (match [x]
      [([_ _ 2] ::m/ints)] :a0
      [([1 1 3] ::m/ints)] :a1
      [([1 2 3] ::m/ints)] :a2
      :else :a3))
  
  ;; ~100ms
  (let [x (int-array [1 2 3])]
    (dotimes [_ 5]
      (time
        (dotimes [_ 1e7]
          (match [^ints x]
            [[_ _ 2]] :a0
            [[1 1 3]] :a1
            [[1 2 3]] :a2)))))

  ;; offsets
  ;; FIXME: needs to account for offset - David
  (let [x (int-array [1 1 2 3])
        o 1]
    (match [x]
      [([_ _ 2] ::m/ints :offset o)] :a0
      [([1 1 3] ::m/ints :offset o)] :a1
      [([1 2 3] ::m/ints :offset o)] :a2))

  ;; 80ms
  (let [x (int-array [1 1 2 3])
        o 1]
    (dotimes [_ 10]
      (time
       (dotimes [_ 1e7]
         (match [x]
           [([_ _ 2] ::m/ints :offset o)] :a0
           [([1 1 3] ::m/ints :offset o)] :a1
           [([1 2 3] ::m/ints :offset o)] :a2)))))

  (do
    (set! *warn-on-reflection* true)
    
    (defmacro asets [a vs]
      `(do
         ~@(map (fn [a b c] (concat a (list b c)))
                (repeat `(aset ~a)) (range (count vs)) vs)
         ~a))
    
    (defn B [l v r]
      (let [^objects o (make-array Object 4)]
        (asets o [:black l v r])))

    (defn R [l v r]
      (let [^objects o (make-array Object 4)]
        (asets o [:red l v r])))
    
    (defn balance-array [node]
      (matchv ::objects [node]
         [(:or [:black [:red [:red a x b] y c] z d]
               [:black [:red a x [:red b y c]] z d]
               [:black a x [:red [:red b y c] z d]]
               [:black a x [:red b y [:red c z d]]])] :balance
         :else :balanced))

    ;; 200ms
    (let [node (B nil nil (R nil nil (R nil nil nil)))]
      (dotimes [_ 10]
        (time
         (dotimes [_ 1e7]
           (balance-array node)))))

    #_(let [node (B nil nil (R nil nil (B nil nil nil)))]
      (balance-array node))
    )
 )














src/main/clojure/clojure/core/scratch.clj
(ns clojure.core.match.test.scratch
 (:require [clojure.core.match :refer [match]]))


(require '[clojure.core.match :refer [match]])



(try
  (cond
    (= y false) (try
                  (cond (= z true) 1
                        :else (throw backtrack))
                  (catch Exception e
                    (if (identical? e backtrack)
                      (throw backtrack)
                      (throw e))))
    
    :else (throw backtrack))
  (catch Exception e
    (if (identical? e backtrack)
      (do
        (try
          (cond
            (= y true) (try (cond
                              (= x false) 2
                              :else (throw backtrack))
                            (catch Exception e
                              (if (identical? e backtrack)
                                (throw backtrack)
                                (throw e))))
            :else (throw backtrack))
          (catch Exception e
            (if (identical? e backtrack)
              (do (try
                    (cond
                      (= z false) 3
                      :else (trhow backtrack))
                    (catch  Exception e
                      (if (identical? e backtrack)
                        (do (try
                              (cond
                                (= z true) 4
                                :else (throw backtrack))
                              (catch Exception e
                                (if (identical? e backtrack)
                                  (throw no-matching-e)
                                  (throw e)))))
                        (throw e)))))
              (throw e)))))
      (throw e))))

;; this should be

(cond (= y false)  (cond (= z true) 1)
      (= y true) (cond (= x false) 2
                       (= z false) 3
                       (= z true) 4))



; loop + jump table (via case) | not cond? 










(let [v [1 2]]
  (clojure.core.match/match [v]
    [] [3 1] :a0
    [([1 a] :as b)] [:a1 a b]))

(let [y '(+ 1 3)
      z true]
  (match [(+ 1 3) y z]
         [false ([1] :seq) true ] 1
         [true true _ ] 2))

(let*
 [y '(+ 1 3) z true]
 (let*
  [ocr-15480 (+ 1 3) y y z z]
  (try
   (if
    (clojure.core/= ocr-15480 false)
    (try
     (if
      (let*
       [and__4467__auto__
        (let*
         [or__4469__auto__ (clojure.core/seq? y)]
         (if
          or__4469__auto__
          or__4469__auto__
          (clojure.core/sequential? y)))]
       (if and__4467__auto__ (clojure.core/seq y) and__4467__auto__))
      (try
       (let*
        [y_head__15485 (clojure.core/first y)]
        (if
         (clojure.core/= y_head__15485 1)
         (try
       ff   (let*
           [y_tail__15486 (clojure.core/rest y)]
           (if
            (clojure.core/empty? y_tail__15486)
            (try
             (if
              (clojure.core/= z true)
              1
              (if :else (throw clojure.core.match/backtrack) nil))
             (catch
              Exception
              e__14590__auto__
              (if
               (clojure.core/identical?
                e__14590__auto__
                clojure.core.match/backtrack
)
               (do (throw clojure.core.match/backtrack))
               (throw e__14590__auto__))))
            (if :else (throw clojure.core.match/backtrack) nil)))
          (catch
           Exception
           e__14590__auto__
           (if
            (clojure.core/identical?
             e__14590__auto__
             clojure.core.match/backtrack)
            (do (throw clojure.core.match/backtrack))
            (throw e__14590__auto__))))
         (if :else (throw clojure.core.match/backtrack) nil)))
       (catch
        Exception
        e__14590__auto__
        (if
         (clojure.core/identical?
          e__14590__auto__
          clojure.core.match/backtrack)
         (do (throw clojure.core.match/backtrack))
         (throw e__14590__auto__))))
      (if :else (throw clojure.core.match/backtrack) nil))
     (catch
      Exception
      e__14590__auto__
      (if
       (clojure.core/identical?
        e__14590__auto__
        clojure.core.match/backtrack)
       (do
        (try
         (if

          (clojure.core/= y true)
          2
          (if :else (throw clojure.core.match/backtrack) nil))
         (catch
          Exception
          e__14590__auto__
          (if
           (clojure.core/identical?
            e__14590__auto__
            clojure.core.match/backtrack)
           (do (throw clojure.core.match/backtrack))
           (throw e__14590__auto__)))))
       (throw e__14590__auto__))))
    (if :else (throw clojure.core.match/backtrack) nil))
   (catch
    Exception
    e__14590__auto__
    (if
     (clojure.core/identical?
      e__14590__auto__
      clojure.core.match/backtrack)
     (do
      (throw
       (new
        java.lang.IllegalArgumentException
        (clojure.core/str
         "No matching clause: "
         ocr-15480
         " "
         y
         " "
         z))))
     (throw e__14590__auto__))))))






(let*
 [ocr-15543 'x ocr-15544 'y ocr-15545 'z]
 (try
  (clojure.core/cond
   (clojure.core/= ocr-15544 false)
   (try
    (clojure.core/cond
     (clojure.core/= ocr-15545 true)
     1
     :else
     (throw clojure.core.match/backtrack))
    (catch
     Exception
     e__14592__auto__
     (if
      (clojure.core/identical?
       e__14592__auto__
       clojure.core.match/backtrack)
      (do (throw clojure.core.match/backtrack))
      (throw e__14592__auto__))))
   :else
   (throw clojure.core.match/backtrack))
  (catch
   Exception
   e__14592__auto__
   (if
    (clojure.core/identical?
     e__14592__auto__
     clojure.core.match/backtrack)
    (do
     (try
      (clojure.core/cond
       (clojure.core/= ocr-15544 true)
       (try
        (clojure.core/cond
         (clojure.core/= ocr-15543 false)
         2
         :else
         (throw clojure.core.match/backtrack))
        (catch
         Exception
         e__14592__auto__
         (if
          (clojure.core/identical?
           e__14592__auto__

           clojure.core.match/backtrack)
          (do (throw clojure.core.match/backtrack))
          (throw e__14592__auto__))))
       :else
       (throw clojure.core.match/backtrack))
      (catch
       Exception
       e__14592__auto__
       (if
        (clojure.core/identical?
         e__14592__auto__
         clojure.core.match/backtrack)
        (do
         (try
          (clojure.core/cond
           (clojure.core/= ocr-15545 false)
           3
           :else
           (throw clojure.core.match/backtrack))
          (catch
           Exception
           e__14592__auto__
           (if
            (clojure.core/identical?
             e__14592__auto__
             clojure.core.match/backtrack)
            (do
             (try
              (clojure.core/cond
               (clojure.core/= ocr-15545 true)
               4
               :else
               (throw clojure.core.match/backtrack))
              (catch
               Exception
               e__14592__auto__
               (if
                (clojure.core/identical?
                 e__14592__auto__
                 clojure.core.match/backtrack)
                (do
                 (throw
                  (java.lang.IllegalArgumentException.
                   (clojure.core/str
                    "No matching clause: "
                    ocr-15543
                    " "
                    ocr-15544
                    " "
                    ocr-15545))))
                (throw e__14592__auto__)))))
            (throw e__14592__auto__)))))
        (throw e__14592__auto__)))))
    (throw e__14592__auto__)))))











;------------------------------




(let*
 [ocr-15543 'x ocr-15544 'y ocr-15545 'z]
 (try
  (clojure.core/cond
   (clojure.core/= y false)
   (try
    (clojure.core/cond
     (clojure.core/= z true)
     1
     :else
     (throw clojure.core.match/backtrack))
    (catch
     Exception
     e__14592__auto__
     (if
      (clojure.core/identical?
       e__14592__auto__
       clojure.core.match/backtrack)
      (do (throw clojure.core.match/backtrack))
      (throw e__14592__auto__))))
   :else
   (throw clojure.core.match/backtrack))
  (catch
   Exception
   e__14592__auto__
   (if
    (clojure.core/identical?
     e__14592__auto__
     clojure.core.match/backtrack)
    (do
     (try
      (clojure.core/cond
       (clojure.core/= y true)
       (try
        (clojure.core/cond
         (clojure.core/= x false)
         2
         :else
         (throw clojure.core.match/backtrack))
        (catch
         Exception
         e__14592__auto__
         (if
          (clojure.core/identical?
           e__14592__auto__

           clojure.core.match/backtrack)
          (do (throw clojure.core.match/backtrack))
          (throw e__14592__auto__))))
       :else
       (throw clojure.core.match/backtrack))
      (catch
       Exception
       e__14592__auto__
       (if
        (clojure.core/identical?
         e__14592__auto__
         clojure.core.match/backtrack)
        (do
         (try
          (clojure.core/cond
           (clojure.core/= z false)
           3
           :else
           (throw clojure.core.match/backtrack))
          (catch
           Exception
           e__14592__auto__
           (if
            (clojure.core/identical?
             e__14592__auto__
             clojure.core.match/backtrack)
            (do
             (try
              (clojure.core/cond
               (clojure.core/= z true)
               4
               :else
               (throw clojure.core.match/backtrack))
              (catch
               Exception
               e__14592__auto__
               (if
                (clojure.core/identical?
                 e__14592__auto__
                 clojure.core.match/backtrack)
                (do
                 (throw
                  (java.lang.IllegalArgumentException.
                   (clojure.core/str
                    "No matching clause: "
                    x
                    " "
                    ocr-15544
                    " "
                    z))))
                (throw e__14592__auto__)))))
            (throw e__14592__auto__)))))
        (throw e__14592__auto__)))))
    (throw e__14592__auto__)))))







(import [java.lang.annotation Retention RetentionPolicy Target ElementType]
        [javax.xml.ws WebServiceRef WebServiceRefs])

(definterface Foo (foo []))

;; annotation on type


[:a1 2 [1 2]]


(defn foo
  "I don't do a whole lot."
  [x]
  (println x "Hello, World!"))
